File: app/api/routes/knowledge.py
"""
知识库管理 API - RAG 向量检索增强
支持文档上传、分块、向量化、语义搜索
"""
import os
from typing import Optional

from fastapi import APIRouter, Body, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.api_response import ApiResponse, ErrCode
from app.services import rag_service

router = APIRouter(prefix="/knowledge", tags=["知识库管理"])

# ── Schemas ──────────────────────────────────────────────────────────


class AddDocumentRequest(BaseModel):
    title: str
    content: str
    file_type: str = "text"  # text/markdown/json


class SearchRequest(BaseModel):
    query: str
    top_k: int = 5
    document_id: Optional[str] = None


class DeleteRequest(BaseModel):
    document_id: str


# ── Endpoints ────────────────────────────────────────────────────────


@router.post("/documents", response_model=ApiResponse)
async def add_document(request: AddDocumentRequest, db: Session = Depends(get_db)):
    """上传文档到知识库，执行向量嵌入"""
    if not request.content.strip():
        return ApiResponse.fail(
            code=ErrCode.VALIDATE_ERROR,
            message="文档内容不能为空"
        )

    result = await rag_service.add_document(
        title=request.title,
        content=request.content,
        file_type=request.file_type,
    )

    if "error" in result:
        return ApiResponse.fail(
            code=ErrCode.SERVER_ERROR,
            message=f"文档上传失败: {result['error']}"
        )

    return ApiResponse.ok(
        data={
            "document_id": result["document_id"],
            "chunk_count": result["chunk_count"],
            "status": result["status"],
        },
        message="文档已添加到知识库"
    )


@router.get("/documents", response_model=ApiResponse)
async def list_documents(
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    """列出知识库中的文档"""
    result = await rag_service.list_documents(limit=limit, offset=offset)
    return ApiResponse.ok(data=result)


@router.get("/search", response_model=ApiResponse)
async def semantic_search(
    q: str = Query(..., description="查询文本"),
    top_k: int = Query(5, ge=1, le=20),
    document_id: Optional[str] = Query(None, description="限定文档ID"),
):
    """向量语义搜索"""
    if not q.strip():
        return ApiResponse.fail(
            code=ErrCode.VALIDATE_ERROR,
            message="查询文本不能为空"
        )

    results = await rag_service.semantic_search(
        query=q,
        top_k=top_k,
        document_id=document_id,
    )

    return ApiResponse.ok(data={"results": results})


@router.delete("/document/{document_id}", response_model=ApiResponse)
async def delete_document(document_id: str, db: Session = Depends(get_db)):
    """删除指定文档"""
    success = await rag_service.delete_document(document_id)

    if not success:
        return ApiResponse.fail(
            code=ErrCode.SERVER_ERROR,
            message="删除失败"
        )

    return ApiResponse.ok(message="文档已删除")


@router.post("/retrieve-context", response_model=ApiResponse)
async def retrieve_context(request: SearchRequest):
    """为对话引擎提供的上下文检索接口"""
    context = await rag_service.retrieve_context(
        query=request.query,
        top_k=request.top_k or 3,
    )
    return ApiResponse.ok(data={"context": context})


__all__ = ["router"]

File: frontend/lib/api/knowledge.ts
/**
 * 知识库（RAG）API — /api/v1/knowledge/*（契约详见 frontend/docs/knowledge-market-recon.md）。
 *
 * 后端事实（UI 不得美化）：
 * - POST /documents 是同步索引：返回即 "completed"，无进度事件/轮询 job，
 *   「解析中」只存在于请求在途层面。
 * - search 的 score 是 FAISS L2 距离：越小越相关，不是归一化相似度。
 * - 后端只收纯文本文档（text/markdown/json）；文件读取在客户端完成。
 * - 删除仅 creator 可操作；先向量后 DB 行；删除比 ≥20% 后端自动压实。
 */
import { apiFetch } from './transport';

/** 后端 ApiResponse<T> 信封（app/models/api_response.py）。 */
interface ApiEnvelope<T> {
  code: string;
  success: boolean;
  message: string;
  data: T | null;
}

export interface KnowledgeDocument {
  id: string;
  title: string;
  file_type: string | null;
  chunk_count: number;
  status: string;
  created_at: string | null;
}

export interface KnowledgeDocumentList {
  total: number;
  items: KnowledgeDocument[];
}

export interface KnowledgeSearchHit {
  /** 分块 id（chk_*） */
  id: string;
  document_id: string;
  title: string;
  /** 分块全文（后端无截断/预览字段） */
  content: string;
  file_type: string | null;
  /** FAISS L2 距离，越小越相关 */
  score: number;
}

export interface AddDocumentResult {
  document_id: string;
  chunk_count: number;
  status: string;
}

/** 后端 MAX_CONTENT_LENGTH（app/api/routes/knowledge.py）。 */
export const KNOWLEDGE_MAX_CONTENT_BYTES = 100 * 1024 * 1024;

export const KNOWLEDGE_FILE_TYPES = ['text', 'markdown', 'json'] as const;
export type KnowledgeFileType = (typeof KNOWLEDGE_FILE_TYPES)[number];

/** 客户端可读入的文本扩展名 → file_type；其余返回 null（诚实拒收）。 */
export function knowledgeFileTypeForFileName(name: string): KnowledgeFileType | null {
  const lower = name.toLowerCase();
  if (lower.endsWith('.md') || lower.endsWith('.markdown')) return 'markdown';
  if (lower.endsWith('.json')) return 'json';
  if (lower.endsWith('.txt')) return 'text';
  return null;
}

export async function listKnowledgeDocuments(
  opts: { limit?: number; offset?: number; signal?: AbortSignal } = {},
): Promise<KnowledgeDocumentList> {
  const params = new URLSearchParams({
    limit: String(opts.limit ?? 20),
    offset: String(opts.offset ?? 0),
  });
  const body = await apiFetch<ApiEnvelope<KnowledgeDocumentList>>(
    `/api/v1/knowledge/documents?${params.toString()}`,
    { signal: opts.signal, label: 'Knowledge docs error' },
  );
  if (!body.success || !body.data) throw new Error(body.message || '无法加载知识库文档');
  return body.data;
}

export async function addKnowledgeDocument(
  req: { title: string; content: string; fileType: KnowledgeFileType },
  opts: { signal?: AbortSignal } = {},
): Promise<AddDocumentResult> {
  const body = await apiFetch<ApiEnvelope<AddDocumentResult>>(
    '/api/v1/knowledge/documents',
    {
      method: 'POST',
      body: { title: req.title, content: req.content, file_type: req.fileType },
      signal: opts.signal,
      label: 'Knowledge document add error',
    },
  );
  if (!body.success || !body.data) throw new Error(body.message || '文档索引失败');
  return body.data;
}

export async function searchKnowledge(
  opts: { q: string; topK?: number; documentId?: string | null; signal?: AbortSignal },
): Promise<KnowledgeSearchHit[]> {
  const params = new URLSearchParams({ q: opts.q, top_k: String(opts.topK ?? 5) });
  if (opts.documentId) params.set('document_id', opts.documentId);
  const body = await apiFetch<ApiEnvelope<{ results: KnowledgeSearchHit[] }>>(
    `/api/v1/knowledge/search?${params.toString()}`,
    { signal: opts.signal, label: 'Knowledge search error' },
  );
  if (!body.success || !body.data) throw new Error(body.message || '检索失败');
  return body.data.results;
}

export async function deleteKnowledgeDocument(
  documentId: string,
  opts: { signal?: AbortSignal } = {},
): Promise<void> {
  const body = await apiFetch<ApiEnvelope<null>>(
    `/api/v1/knowledge/document/${encodeURIComponent(documentId)}`,
    { method: 'DELETE', signal: opts.signal, label: 'Knowledge document delete error' },
  );
  if (!body.success) throw new Error(body.message || '删除失败或无权访问该文档');
}
